Add a grid for the flow past a cylinder benchmark.

drwells opened the pull request:
Fixes #6675. I duplicated the grid found here:
http://www.featflow.de/en/benchmarks/cfdbenchmarking/flow/dfg_benchmark1_re20.html

Here is the 3D grid:
![channel_with_cylinder_3d](https://user-images.githubusercontent.com/3917035/42116613-0a10d908-7bc6-11e8-89d5-e07b5abd6357.png)

This could be part of the tutorial on how to use the TFI manifold class.

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -1910,6 +1910,283 @@ namespace GridGenerator
+  {
+    Triangulation<2> tria_2;
+    internal::channel_with_cylinder(tria_2);
+    extrude_triangulation(tria_2, 5, 0.41, tria);
```
this function is so handy :-)

drwells commented on source/grid/grid_generator.cc:
```diff
@@ -1910,6 +1910,283 @@ namespace GridGenerator
+    namespace
+    {
+      // Implementation of channel_with_cylinder: sets the cells adjacent to
+      // the cylinder's material ids to 2. Keeping this set is useful in 3D,
```
I watched a technical talk about a month ago where the speaker noted that 'get' and 'set' have the longest definitions in the Oxford English dictionary.

What I meant (and this is, in retrospect, not clear at all) was 'Keeping *the material ids of the cells adjacent to the cylinder* set *to 2* is useful in 3D'. I'll reword this into something better.

bangerth replied:
Ha, how funny with the OED. And thanks for the clarification!

drwells commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -304,6 +304,59 @@ namespace GridGenerator
+   * or CylindricalManifold (in 3D) with manifold id $0$, a
+   * TransfiniteInterpolationManifold with manifold id $1$, and a FlatManifold
+   * everywhere else. The cell faces on the cylinder have a manifold id of
+   * $0$, while the cells themselves have a manifold id of $1$. Put another
```
What I wrote is vague: I meant that the cells (the codim 0 objects, not faces, lines, etc) adjacent to the cylinder have a manifold id of 1. I'll rewrite this.

drwells commented on source/grid/grid_generator.cc:
```diff
@@ -1910,6 +1910,283 @@ namespace GridGenerator
+        for (const auto cell : cylinder_tria.active_cell_iterators())
+          cell->set_material_id(2);
+
+        auto minimal_line_length = [](const Triangulation<2> &t) -> double {
```
We have `cell->enclosing_ball()` but I don't see the opposite function (i.e., `cell->enscribed_ball()`). Would these be somewhere besides `tria_accessor.h`?

drwells replied:
I prefer `line` since this loops over the lines of the cell. Why do you prefer edge?

bangerth replied:
The function I was thinking of is `TriaAccessor:: minimum_vertex_distance()`.

edge vs line -- I don't feel strongly enough about to try to explain the difference in preference ;-)

drwells commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -304,6 +304,59 @@ namespace GridGenerator
+   * A channel with an embedded cylinder. This is a common benchmark for
+   * Navier-Stokes solvers. The geometry consists of a channel of size $[0,
+   * 2.2] \times [0, 0.41] \times [0, 0.41]$ (where the $z$ dimension is
+   * ommited in 2D) with a cylinder, parallel to the $z$ axis with diameter
```
Fixed

kronbichler commented:
I have one comment, though: Depending on the Reynolds number one might want to refine the mesh close to the cylinder. In one of our papers over here, https://www.sciencedirect.com/science/article/pii/S0021999117306915 we have computed accurate results for the Reynolds number of 100, where we found best results when there were two more layers of circular-shaped elements, see Fig 11 in that paper for the mesh we used, reproduced here, so ignore the upper part of the figure:

![cylinder](https://user-images.githubusercontent.com/8276200/42277333-4b2e3d5c-7f97-11e8-845a-eb21bd560384.png)

We could add those here by an argument `n_additional_circular_cells` and some grid stretching factor as you see in this figure. What do you think?

drwells commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -304,6 +304,59 @@ namespace GridGenerator
+   * @image html channel_with_cylinder_3d.png
+   *
+   * The resulting Triangulation uses three manifolds: a PolarManifold (in 2D)
+   * or CylindricalManifold (in 3D) with manifold id $0$, a
```
I am not sure exactly what you have in mind here; I just added a generic 'for more information click here' sentence in both places. If you have something better let me know.

bangerth replied:
That works. I meant to suggest just using `@ref GlossManifoldId "manifold id"` where you currently just say "manifold id" in the text.

drwells commented on source/grid/grid_generator.cc:
```diff
@@ -1910,6 +1910,283 @@ namespace GridGenerator
+            if ((cell->center() - Point<2>(0.2, 0.2)).norm() < 0.15)
+              cells_to_remove.insert(cell);
+
+            if (cylinder_triangulation_offset == Point<2>())
```
Fixed

drwells commented on source/grid/grid_generator.cc:
```diff
@@ -1910,6 +1910,283 @@ namespace GridGenerator
+                                                  Point<2>(2.2, 0.41));
+        std::set<Triangulation<2>::active_cell_iterator> cells_to_remove;
+        Tensor<1, 2> cylinder_triangulation_offset;
+        for (const auto &cell : bulk_triangulation.active_cell_iterators())
```
I added a comment and a picture.

drwells commented:
@bangerth I believe I addressed all comments.

@kronbichler I read the arxiv version of the manuscript. I think I know how to add more cells this way (use anisotropic refinement and copy the result to level 0) but its not clear to me how to do the stretching. Do you have a reference for that part?

drwells commented:
Here is the new picture that @bangerth requested:

![channel_with_cylinder_2d_manifolds](https://user-images.githubusercontent.com/3917035/42285050-5435d4f2-7f7c-11e8-826e-1aeaada872bd.png)

kronbichler commented:
@drwells If you do it with anisotropic refinement, it probably gets challenging. What I was thinking about is to simply add one (or more) shell (annulus) geometries at the interior. If you do that, you only need to make the inner radius of the current TFI part (which is really nice, by the way) bigger by the amount. If we want to have a skewness, the typical setup is to do this via a tanh function:
If we have to cover a radius between `0.1` and `0.15`, then I would compute intermediate radii by `0.1 + 0.05*(1 + std::tanh(skewness * i / n_radii) / std::tanh(skewness)`. In this formula, a skewness of 0 would be a linear distribution (one needs to use a special case to avoid `0/0`, but that's the general procedure at least in the CFD community.

drwells commented:
You're right; its easier to merge shells than do accounting with anisotropic refinement.

I'll add two more arguments: one for the number of shells and one for the skewness. I'll report back when things are working.

kronbichler commented:
Oh and thanks a lot for the effort, I remember it took us several days to get this right the first time around.

davydden commented on source/grid/grid_generator.cc:
```diff
@@ -1910,6 +1910,313 @@ namespace GridGenerator
+        // a result the vertices might not be lined up. Get around this by
+        // deleting duplicated vertices with a very low numerical tolerance.
+        const double tolerance =
+          std::min(minimal_line_length(tria_without_cylinder),
```
since you hard-code dimension, don't you know which tolerance works in advance?

drwells replied:
Yes, and I hard-code other tolerances in other places. Its not consistent. Would you suggest doing all one or the other?

davydden replied:
I would probably hardcode it here as well to make code shorter.

davydden commented on source/grid/grid_generator.cc:
```diff
@@ -1910,6 +1910,313 @@ namespace GridGenerator
+          if (cell->material_id() != 2)
+            cell->set_all_manifold_ids(numbers::flat_manifold_id);
+
+        // We need to calculate the current center so that we can move it later:
```
maybe move the whole center shift part to `hyper_cube_with_cylindrical_hole` via additional parameter? Can be useful elsewhere.

davydden commented:
p.s. I think it would also be nice to have an option higher spatial resolution across the radius in `hyper_cube_with_cylindrical_hole()`, as opposed to hard-coding this in this generator function (if I understand correctly @kronbichler mesh correctly).

masterleinad commented:
The failure of the `clang` build seems to be spurious:
```
FAILED: cd /home/bob/build-clang/source/dofs && /home/bob/build-clang/bin/expand_instantiations /home/bob/build-clang/share/deal.II/template-arguments < /home/bob/source/source/dofs/block_info.inst.in > /home/bob/build-clang/source/dofs/block_info.inst.tmp && /usr/bin/cmake -E rename /home/bob/build-clang/source/dofs/block_info.inst.tmp /home/bob/build-clang/source/dofs/block_info.inst
Invalid instantiation list: missing 'for'
ninja: build stopped: subcommand failed.
```

drwells commented:
I figured out how to add the shells; I will try to finish this up in the next few days.

drwells commented:
I will finish this once #6924 is merged: we should use that geometry directly here so that users can generate the grid @kronbichler proposed with one function call.

drwells commented:
This is now done, but it depends on #7049 (i.e., it won't compile without that PR).

drwells commented:
Here are the new pictures:

![channel_with_cylinder_2d](https://user-images.githubusercontent.com/3917035/44012102-c2d47f76-9e8a-11e8-9fe3-ff78f18ea603.png)

![channel_with_cylinder_3d](https://user-images.githubusercontent.com/3917035/44012099-bf562f7a-9e8a-11e8-98ab-118f37756fe3.png)

![channel_with_cylinder_2d_manifolds](https://user-images.githubusercontent.com/3917035/44012105-c5e574ae-9e8a-11e8-9a42-5335e2d1b04b.png)

kronbichler commented:
I think this looks great. Thanks for the effort @drwells !

drwells commented:
This is now rebased against `master`.

drwells commented:
/run-tests

kronbichler commented:
Since there were no further comments, I think this can be merged. Thanks for the nice work @drwells
